File: lib/dump.ts
// Dump orchestration: given a (live or mocked) CoveoClient and resolved options,
// fetch each configured type, write one JSON per article + the per-type field
// catalogue, and emit _index.json. This is the per-type loop extracted VERBATIM
// from cmd/dump.ts so it can be driven offline in tests with an injected
// CoveoClient. cmd/dump.ts keeps arg parsing / config loading / client
// construction and calls dumpTypes(); the data-affecting logic is unchanged.

import { type Logger, NULL_LOGGER } from "./logger.ts";
import { normalizeType, type TypeConfig } from "./config/types.ts";
import { CoveoClient, type CoveoResult } from "./coveo/client.ts";
import { dateAq, modMsOf } from "./coveo/dates.ts";
import { fetchTypeSince } from "./coveo/paging.ts";
import {
  type CatalogueEntry,
  flattenFieldsSafe,
  splitEntry,
  updateCatalogue,
  writeCatalogue,
} from "./coveo/fields.ts";
import { idOf, sanitizeName } from "./fsutil.ts";
import { makeProgress } from "./progress.ts";
import { sha256 } from "./track/hashing.ts";
import type { Changelog } from "./changelog.ts";

// DB key for an article: matches the (document_type, id) primary key in articles.db.
export function dbKey(documentType: string, id: string): string {
  return `${documentType} ${id}`;
}

export interface TypeStatus {
  typeKey: string;
  documentType: string;
  dir: string;
  status: "ok" | "partial" | "failed";
  expected: number | null;
  fetched: number;
  written: number;
  /** incremental mode: unchanged articles left untouched (not rewritten). */
  skipped: number;
  writeErrors: number;
  error?: string;
}

export interface DumpTypesOpts {
  /** type key -> config (documentType + keep-lists); iterated by typeKeys order. */
  typeConfigs: Record<string, Partial<TypeConfig>>;
  /** type keys to dump (already filtered/validated by the caller). */
  typeKeys: string[];
  /** field-name -> catalogue description. */
  descriptions: Record<string, string>;
  outDir: string;
  allTime: boolean;
  /** _index.json "mode" string: "all" or "days=N" (matches dump_articles.ts). */
  mode: string;
  /** epoch ms lower bound (ignored when allTime). */
  cutoffMs: number;
  /** epoch ms upper bound. */
  endMs: number;
  /** capturedAt stamp (epoch ms) — one value across the whole run. */
  nowMs: number;
  pageSize: number;
  limit: number;
  configPath: string;
  logger?: Logger;
  /** incremental mode: skip rewriting articles whose metadata_hash is unchanged. */
  incremental?: boolean;
  /** dbKey(documentType,id) -> metadata_hash from the DB (loaded by the caller). */
  priorHashes?: Map<string, string>;
  /** optional changelog sink (records added/edited). */
  changelog?: Changelog;
  /** preview: classify + record changelog but write no files (no article/catalogue/_index). */
  dryRun?: boolean;
}

export interface DumpTypesResult {
  manifest: TypeStatus[];
  indexPath: string;
  total: number;
  /** typeKey -> set of article ids present in Coveo this run (for deletion reconcile). */
  currentIds: Map<string, Set<string>>;
}

// Run the per-type dump loop and write _index.json. Returns the manifest so the
// caller can compute the exit code (1 if any type FAILED).
export async function dumpTypes(
  client: CoveoClient,
  opts: DumpTypesOpts,
): Promise<DumpTypesResult> {
  const logger = opts.logger ?? NULL_LOGGER;
  const {
    typeConfigs,
    typeKeys,
    descriptions,
    outDir,
    allTime,
    mode,
    cutoffMs,
    endMs,
    nowMs,
    pageSize,
    limit,
    configPath,
    incremental,
    priorHashes,
    changelog,
    dryRun,
  } = opts;

  if (!dryRun) await Deno.mkdir(outDir, { recursive: true });

  const manifest: TypeStatus[] = [];
  const currentIds = new Map<string, Set<string>>();

  for (const typeKey of typeKeys) {
    const cfg: TypeConfig = normalizeType({
      documentType: typeConfigs[typeKey]?.documentType,
      metadata: typeConfigs[typeKey]?.metadata,
      content: typeConfigs[typeKey]?.content,
    });
    const dir = sanitizeName(typeKey);
    if (!cfg.documentType) {
      logger.warn(`Skipping "${typeKey}": no documentType in config`);
      manifest.push({
        typeKey,
        documentType: "",
        dir,
        status: "failed",
        expected: null,
        fetched: 0,
        written: 0,
        skipped: 0,
        writeErrors: 0,
        error: "no documentType in config",
      });
      continue;
    }

    const st: TypeStatus = {
      typeKey,
      documentType: cfg.documentType,
      dir,
      status: "ok",
      expected: null,
      fetched: 0,
      written: 0,
      skipped: 0,
      writeErrors: 0,
    };
    const idSet = new Set<string>();
    currentIds.set(typeKey, idSet);
    const progress = makeProgress(logger);
    try {
      // Server-side count over the window — the target to validate against.
      const expectAq = allTime
        ? `@f5_document_type=="${cfg.documentType}"`
        : `@f5_document_type=="${cfg.documentType}" ${dateAq(cutoffMs, endMs)}`.trim();
      st.expected = await client.getCount(expectAq);

      progress.start(typeKey, st.expected ?? undefined);
      const results = await fetchTypeSince(
        client,
        cfg.documentType,
        cutoffMs,
        endMs,
        pageSize,
        limit,
        (n) => progress.update(n),
        !allTime,
      );
      st.fetched = results.length;

      const typeDir = `${outDir}/${dir}`;
      if (!dryRun) await Deno.mkdir(typeDir, { recursive: true });

      const catalogue = new Map<string, CatalogueEntry>();
      const seenIds = new Map<string, number>();

      for (const r of results) {
        const fields = flattenFieldsSafe(r);
        updateCatalogue(catalogue, fields, descriptions);

        const { metadata, content } = splitEntry(fields, cfg);
        const raw = (r.raw as CoveoResult) ?? {};

        let id = idOf(r);
        const n = (seenIds.get(id) ?? 0) + 1;
        seenIds.set(id, n);
        if (n > 1) id = `${id}__${n}`;

        idSet.add(id); // present in Coveo this run (for deletion reconcile)

        const modMs = modMsOf(raw);
        const title = (r.title as string) ?? "";

        // Compare against the DB's stored metadata_hash when we have one (incremental
        // skip and/or a changelog need it). Incremental: an unchanged article is left
        // untouched (the existing, possibly-enriched file stays). Non-incremental with
        // a changelog: we still rewrite, but only log genuinely added/edited articles.
        if (priorHashes && (incremental || changelog)) {
          const mh = await sha256(metadata);
          const prior = priorHashes.get(dbKey(cfg.documentType, id));
          const unchanged = prior !== undefined && prior === mh;
          if (unchanged) {
            if (incremental) {
              st.skipped++;
              continue; // skip write; enrich (resumable) won't touch it either
            }
            // non-incremental: rewritten but not a change — don't log it.
          } else {
            changelog?.record({
              op: prior === undefined ? "added" : "edited",
              documentType: cfg.documentType,
              id,
              title,
              hashOld: prior,
              hashNew: mh,
              source: "dump",
            });
          }
        }

        const entry = {
          id,
          documentType: cfg.documentType,
          title,
          link: (r.clickUri as string) ?? (raw.clickableuri as string) ?? "",
          modifiedMs: modMs ?? null,
          modified: modMs ? new Date(modMs).toISOString() : null,
          capturedAt: new Date(nowMs).toISOString(),
          metadata,
          content,
        };
        try {
          if (!dryRun) {
            await Deno.writeTextFile(`${typeDir}/${id}.json`, JSON.stringify(entry, null, 2));
          }
          st.written++;
        } catch (e) {
          st.writeErrors++;
          if (st.writeErrors <= 3) logger.warn(`write failed for ${id}: ${(e as Error).message}`);
        }
      }

      if (!dryRun) {
        await writeCatalogue(typeDir, typeKey, cfg.documentType, catalogue, results.length, cfg);
      }

      // Undercount=partial only under --all. In incremental mode unchanged articles
      // are present-but-skipped, so completeness = written + skipped.
      const present = st.written + st.skipped;
      const undercount = allTime && st.expected !== null && limit === Infinity &&
        present < st.expected;
      if (st.writeErrors > 0 || undercount) st.status = "partial";

      const flag = st.status === "ok" ? "" : `  [${st.status.toUpperCase()}]`;
      const exp = st.expected !== null ? `/${st.expected}` : "";
      const skip = st.skipped ? ` (${st.skipped} unchanged)` : "";
      progress.done(
        `${st.written}${exp} written${skip} article${
          st.written === 1 ? "" : "s"
        } -> ${typeDir}/${flag}`,
      );
    } catch (e) {
      st.status = "failed";
      st.error = (e as Error).message;
      progress.done(`FAILED: ${st.error}`);
    }
    manifest.push(st);
  }

  const failed = manifest.filter((m) => m.status === "failed");
  const partial = manifest.filter((m) => m.status === "partial");
  const total = manifest.reduce((a, m) => a + m.written, 0);

  const indexPath = `${outDir}/_index.json`;
  if (!dryRun) {
    await Deno.writeTextFile(
      indexPath,
      JSON.stringify(
        {
          mode,
          cutoff: new Date(cutoffMs).toISOString(),
          generatedAt: new Date(nowMs).toISOString(),
          config: configPath,
          totalArticles: total,
          counts: {
            types: manifest.length,
            ok: manifest.filter((m) => m.status === "ok").length,
            partial: partial.length,
            failed: failed.length,
          },
          types: manifest,
        },
        null,
        2,
      ),
    );
  }

  return { manifest, indexPath, total, currentIds };
}

File: python/f5kb/lib/dump.py
"""Dump orchestration: fetch types via Coveo, write articles + catalogue + _index."""

from __future__ import annotations

import datetime
import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from f5kb.lib.logger import Logger, NULL_LOGGER
from f5kb.lib.fsutil import id_of, sanitize_name, write_json
from f5kb.lib.progress import make_progress
from f5kb.lib.staging import (
    PendingEntry,
    archive_replaced,
    live_article,
    now_stamp,
    pending_path,
)
from f5kb.config.types import TypeConfig, normalize_type
from f5kb.coveo.client import CoveoClient
from f5kb.coveo.dates import date_aq, mod_ms_of
from f5kb.coveo.paging import fetch_type_since
from f5kb.coveo.fields import (
    CatalogueEntry,
    flatten_fields_safe,
    split_entry,
    update_catalogue,
    write_catalogue,
)
from f5kb.track.hashing import sha256_obj


def db_key(document_type: str, id: str) -> str:
    """Exact match for load_hash_index key format: '<document_type> <id>'."""
    return f"{document_type} {id}"


@dataclass
class TypeStatus:
    type_key: str
    document_type: str
    dir: str
    status: str  # "ok"|"partial"|"failed"
    expected: int | None
    fetched: int
    written: int
    skipped: int
    staged: int
    replaced: int
    write_errors: int
    error: str | None = None


@dataclass
class DumpTypesResult:
    manifest: list[TypeStatus]
    index_path: str
    total: int
    current_ids: dict[str, set[str]]  # typeKey -> set of ids in Coveo this run
    pending: list[PendingEntry]


def dump_types(
    client: CoveoClient,
    *,
    type_configs: dict[str, dict],
    type_keys: list[str],
    descriptions: dict[str, str],
    out_dir: str,
    all_time: bool,
    mode: str,
    cutoff_ms: int,
    end_ms: int,
    now_ms: int,
    page_size: int,
    limit: int,
    config_path: str,
    logger: Logger = NULL_LOGGER,
    incremental: bool = False,
    prior_hashes: dict[str, str] | None = None,
    changelog=None,
    dry_run: bool = False,
    approval: bool = False,
    archive_on_overwrite: bool = False,
) -> DumpTypesResult:
    log = logger
    if not dry_run:
        Path(out_dir).mkdir(parents=True, exist_ok=True)

    manifest: list[TypeStatus] = []
    current_ids: dict[str, set[str]] = {}
    pending: list[PendingEntry] = []
    stamp = now_stamp(now_ms)
    captured_iso = datetime.datetime.fromtimestamp(now_ms / 1000, tz=datetime.timezone.utc).isoformat().replace("+00:00", "Z")

    for type_key in type_keys:
        raw_cfg = type_configs.get(type_key) or {}
        cfg = normalize_type({
            "documentType": raw_cfg.get("documentType"),
            "metadata": raw_cfg.get("metadata"),
            "content": raw_cfg.get("content"),
        })
        dir_name = sanitize_name(type_key)
        if not cfg.document_type:
            log.warn(f'Skipping "{type_key}": no documentType in config')
            manifest.append(TypeStatus(
                type_key=type_key, document_type="", dir=dir_name, status="failed",
                expected=None, fetched=0, written=0, skipped=0, staged=0, replaced=0,
                write_errors=0, error="no documentType in config",
            ))
            continue

        st = TypeStatus(
            type_key=type_key, document_type=cfg.document_type, dir=dir_name,
            status="ok", expected=None, fetched=0, written=0, skipped=0,
            staged=0, replaced=0, write_errors=0,
        )
        id_set: set[str] = set()
        current_ids[type_key] = id_set
        progress = make_progress(log)

        try:
            expect_aq = (
                f'@f5_document_type=="{cfg.document_type}"'
                if all_time
                else f'@f5_document_type=="{cfg.document_type}" {date_aq(cutoff_ms, end_ms)}'.strip()
            )
            st.expected = client.get_count(expect_aq)
            progress.start(type_key, st.expected)

            results = fetch_type_since(
                client, cfg.document_type, cutoff_ms, end_ms, page_size, limit,
                progress_cb=lambda n: progress.update(n),
                use_date_window=not all_time,
            )
            st.fetched = len(results)

            type_dir = str(Path(out_dir) / dir_name)
            if not dry_run:
                Path(type_dir).mkdir(parents=True, exist_ok=True)

            catalogue: dict[str, CatalogueEntry] = {}
            seen_ids: dict[str, int] = {}

            for r in results:
                fields = flatten_fields_safe(r)
                update_catalogue(catalogue, fields, descriptions)

                metadata, content = split_entry(fields, cfg)
                raw = r.get("raw") or {}
                art_id = id_of(r)

                n = seen_ids.get(art_id, 0) + 1
                seen_ids[art_id] = n
                if n > 1:
                    art_id = f"{art_id}__{n}"

                id_set.add(art_id)

                mod_ms = mod_ms_of(raw)
                title = r.get("title") or ""
                entry = {
                    "id": art_id,
                    "documentType": cfg.document_type,
                    "title": title,
                    "link": r.get("clickUri") or raw.get("clickableuri") or "",
                    "modifiedMs": mod_ms,
                    "modified": datetime.datetime.fromtimestamp(mod_ms / 1000, tz=datetime.timezone.utc).isoformat().replace("+00:00", "Z") if mod_ms else None,
                    "capturedAt": captured_iso,
                    "metadata": metadata,
                    "content": content,
                }

                gated = approval or archive_on_overwrite
                classify = incremental or changelog or gated
                unchanged = False
                is_edited = False
                is_new = False
                mh: str | None = None
                prior: str | None = None

                if classify:
                    mh = sha256_obj(metadata)
                    prior = (prior_hashes or {}).get(db_key(cfg.document_type, art_id))
                    if prior is None and gated:
                        lf = live_article(out_dir, dir_name, art_id)
                        if lf:
                            prior = sha256_obj(lf.get("metadata") or {})
                    unchanged = prior is not None and prior == mh
                    is_edited = prior is not None and prior != mh
                    is_new = prior is None

                if unchanged and (incremental or gated):
                    st.skipped += 1
                    continue

                if is_new and changelog:
                    changelog.record("added", cfg.document_type, art_id, title=title, hashNew=mh, source="dump")

                if is_edited and approval:
                    if not dry_run:
                        pp = pending_path(out_dir, dir_name, art_id)
                        Path(pp).parent.mkdir(parents=True, exist_ok=True)
                        Path(pp).write_text(json.dumps(entry, indent=2))
                    st.staged += 1
                    pending.append(PendingEntry(
                        type_key=dir_name, id=art_id, title=title,
                        op="edited", changed=["metadata"], source="dump",
                        hash_old=prior, hash_new=mh, staged_at=captured_iso,
                    ))
                    continue

                if is_edited:
                    if changelog:
                        changelog.record("edited", cfg.document_type, art_id, title=title, hashOld=prior, hashNew=mh, source="dump")
                    if archive_on_overwrite and not dry_run:
                        arch = archive_replaced(out_dir, dir_name, art_id, stamp)
                        if arch:
                            st.replaced += 1

                try:
                    if not dry_run:
                        Path(type_dir, f"{art_id}.json").write_text(json.dumps(entry, indent=2))
                    st.written += 1
                except Exception as e:
                    st.write_errors += 1
                    if st.write_errors <= 3:
                        log.warn(f"write failed for {art_id}: {e}")

            if not dry_run:
                write_catalogue(type_dir, type_key, cfg.document_type, catalogue, len(results), cfg)

            present = st.written + st.skipped + st.staged
            undercount = all_time and st.expected is not None and limit == 0 and present < st.expected
            if st.write_errors > 0 or undercount:
                st.status = "partial"

            exp = f"/{st.expected}" if st.expected is not None else ""
            skip = f" ({st.skipped} unchanged)" if st.skipped else ""
            stg = f" ({st.staged} staged for approval)" if st.staged else ""
            flag = f"  [{st.status.upper()}]" if st.status != "ok" else ""
            progress.done(f"{st.written}{exp} written{skip}{stg} article{'s' if st.written != 1 else ''} -> {type_dir}/{flag}")

        except Exception as e:
            st.status = "failed"
            st.error = str(e)
            progress.done(f"FAILED: {st.error}")

        manifest.append(st)

    total = sum(m.written for m in manifest)
    index_path = str(Path(out_dir) / "_index.json")
    if not dry_run:
        Path(index_path).write_text(json.dumps({
            "mode": mode,
            "cutoff": datetime.datetime.fromtimestamp(cutoff_ms / 1000, tz=datetime.timezone.utc).isoformat().replace("+00:00", "Z"),
            "generatedAt": datetime.datetime.fromtimestamp(now_ms / 1000, tz=datetime.timezone.utc).isoformat().replace("+00:00", "Z"),
            "config": config_path,
            "totalArticles": total,
            "counts": {
                "types": len(manifest),
                "ok": sum(1 for m in manifest if m.status == "ok"),
                "partial": sum(1 for m in manifest if m.status == "partial"),
                "failed": sum(1 for m in manifest if m.status == "failed"),
            },
            "types": [
                {
                    "typeKey": m.type_key,
                    "documentType": m.document_type,
                    "dir": m.dir,
                    "status": m.status,
                    "expected": m.expected,
                    "fetched": m.fetched,
                    "written": m.written,
                    "skipped": m.skipped,
                    "staged": m.staged,
                    "replaced": m.replaced,
                    "writeErrors": m.write_errors,
                    **({"error": m.error} if m.error else {}),
                }
                for m in manifest
            ],
        }, indent=2))

    return DumpTypesResult(manifest=manifest, index_path=index_path, total=total, current_ids=current_ids, pending=pending)
